pkg/kubelet/winstats/perfcounter_nodestats.go

```diff
@@ -82,62 +93,61 @@ func (p *perfCounterNodeStatsClient) getMachineInfo() (*cadvisorapi.MachineInfo,
 
 	return &cadvisorapi.MachineInfo{
 		NumCores:       runtime.NumCPU(),
-		MemoryCapacity: p.memoryPhysicalCapacityBytes,
+		MemoryCapacity: p.nodeInfo.memoryPhysicalCapacityBytes,
 		MachineID:      hostname,
 	}, nil
 }
+
 func (p *perfCounterNodeStatsClient) getVersionInfo() (*cadvisorapi.VersionInfo, error) {
 	return &cadvisorapi.VersionInfo{
-		KernelVersion: p.kernelVersion,
+		KernelVersion: p.nodeInfo.kernelVersion,
 	}, nil
 }
 
-func (p *perfCounterNodeStatsClient) getNodeStats() (nodeStats, error) {
+func (p *perfCounterNodeStatsClient) getNodeMetrics() (nodeMetrics, error) {
 	p.mu.RLock()
 	defer p.mu.RUnlock()
-	return p.nodeStats, nil
+	return p.nodeMetrics, nil
 }
 
-func (p *perfCounterNodeStatsClient) startNodeMonitoring(cpuChan, memWorkingSetChan, memCommittedBytesChan chan metric) {
-	for {
-		select {
-		case cpu := <-cpuChan:
-			p.updateCPU(cpu)
-		case mWorkingSet := <-memWorkingSetChan:
-			p.updateMemoryWorkingSet(mWorkingSet)
-		case mCommittedBytes := <-memCommittedBytesChan:
-			p.updateMemoryCommittedBytes(mCommittedBytes)
-		}
-		p.lastUpdatedTime = time.Now()
-	}
+func (p *perfCounterNodeStatsClient) getNodeInfo() nodeInfo {
+	return p.nodeInfo
 }
-func (p *perfCounterNodeStatsClient) updateCPU(cpu metric) {
+
+func (p *perfCounterNodeStatsClient) collectMetricsData(cpuCounter, memWorkingSetCounter, memCommittedBytesCounter *perfCounter) {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	cpuCores := runtime.NumCPU()
-	// This converts perf counter data which is cpu percentage for all cores into nanoseconds.
-	// The formula is (cpuPercentage / 100.0) * #cores * 1e+9 (nano seconds). More info here:
-	// https://github.com/kubernetes/heapster/issues/650
-	newValue := p.cpuUsageCoreNanoSeconds.Value + uint64((float64(cpu.Value)/100.0)*float64(cpuCores)*1000000000)
+	cpuValue, err := cpuCounter.getData()
+	if err != nil {
+		return
+	}
 
-	p.cpuUsageCoreNanoSeconds = metric{
-		Name:      cpu.Name,
-		Value:     newValue,
-		Timestamp: cpu.Timestamp,
+	memWorkingSetValue, err := memWorkingSetCounter.getData()
+	if err != nil {
+		return
 	}
-}
 
-func (p *perfCounterNodeStatsClient) updateMemoryWorkingSet(mWorkingSet metric) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	p.memoryPrivWorkingSetBytes = mWorkingSet
+	memCommittedBytesValue, err := memCommittedBytesCounter.getData()
+	if err != nil {
+		return
+	}
+
```

Comment:
Minimize the critical section by moving the lock acquiring here.

Reply:
Fixed, thanks for the suggestion.